client/src/lib/actions/profile_setup.ts

```diff
@@ -21,63 +34,171 @@ export type ProfileSetupState = {
-
-  // Retrieve the authenticated user to get the user ID
-  const {
-    data: { user },
-    error: userError,
-  } = await supabase.auth.getUser();
-  console.log("User: ", user);
-
+  const { data: {user}, error: userError } =  await supabase.auth.getUser();
   if (userError) {
-    console.log("Error retrieving user:", userError.message);
-    return { message: "Error retrieving user information" };
+      debug("ERROR: " + userError);
+      return null;
   }
-
-  // Ensure we have a valid user ID
+  
   const userId = user?.id;
-  if (!userId) {
-    return { message: "User not authenticated" };
+  
+  if(userId) {
+    return userId;
   }
+  return null;
+} // end of getUserId
 
-  // TODO: Finish implementation to insert a profile
-  // update the profile in Supabase
-  const { error } = await supabase.from("profile").insert({
-    user_id: userId,
-    username,
-    school,
-    avatar,
-  });
+//getProfileForUser gets profile for currently auth'd user
+export async function getProfileForUser() : Promise<Json|null>{
+  debug("start of getProfileForUser() function");
+  const user_id = await getUserId();
+  if(!user_id) {
+    debug("No user_id: GetUserId() returned null");
+    return null;
+  }
+  const supabase = createSupabaseServer();
+    
+  //query for profile by user_id
+  const { data, error } = await supabase
+    .from("profile")
+    .select('*')
+    .eq("user_id", user_id);
+
+    if (error) {
+      debug("ERROR: unable to fetch profile by userid");
+      debug("error: "+ error);
+      return null;
+    }
+
+    const profileDataStr = JSON.stringify(data, null, 2);
+    const profileJson = JSON.parse(profileDataStr);
+    
+    if (profileJson && profileJson[0]) {
+      return profileJson[0];
+    }
+    else {
+      return null
+    }
+} // end of getProfileForUser
+
+export async function getProfileByUserName(username: string) : Promise<Json|null>{
+
+  debug("start of getProfileByUserName() function");
+  const supabase = createSupabaseServer();
+    
+  //query for profile by username
+  const { data, error } = await supabase
+    .from("profile")
+    .select('*')
+    .eq("username", username);
+
+    if (error) {
+      debug("ERROR: retrieving profile by username: "+ error);
+      return null;
+    }
 
-  if (error) {
-    console.log("Error occurred creating profile:", error);
-    return { message: "Database Error" };
+    const profileDataStr = JSON.stringify(data, null, 2);
+    const profileJson = JSON.parse(profileDataStr);
+
+    if (profileJson && profileJson[0]) {
+      return profileJson[0];
+    }
+    else {
+      return null
+    }
+} // end of getProfileByUserName
+
+export async function createProfile( prevState: ProfileSetupState, formData: FormData ): Promise<ProfileSetupState> {
+  if (await userHasExistingProfile()) {
+    debug("USER has an existing profile.")
+    return {
+      errors: {
+        school: ["ERROR: User has an existing profile"],
```

Comment:
The user shouldn't be able to access the /profile_setup page when they have a profile already. So you don't have to do this check.